New Ruby Fathead

kotoshenya opened the pull request:
Scripts to download and extract the [ruby-doc.org](http://ruby-doc.org) Ruby core documentation from http://ruby-doc.org/downloads/ruby_2_3_1_core_rdocs.tgz (configured in `data.url`), scrape it for method and class/module documentation summaries, and write those to `output.txt`.  The actual parsing and formatting is done by `parse.rb`.

This should provide instant answers for queries such as

    ruby array length

and

    string ruby

### Shell scripts

- `lib/fathead/ruby/fetch.sh`: download and extract the ruby-doc.org source.
- `lib/fathead/ruby/parse.sh`: follow each class/module link from the ruby-doc.org index page, adding the class/module description and any class/instance method descriptions to `output.txt`.

### Notes on the current implementation

- URLs are links to pages under http://ruby-doc.org/core/ (e.g., [Enumerable.html](http://ruby-doc.org/core/Enumerable.html)), which redirect to docs for the latest version of ruby (http://ruby-doc.org/core-2.3.1/ as of this commit.)
- All methods are namespaced, including those on Object and Kernel.
- Cross-references are not handled (links in the abstract are stripped).
- Scraping the documentation takes about 10 seconds on my machine.

### Sample of `output.txt`

    Array	A			Ruby classes							Arrays are ordered, integer-indexed collections of any object.	http://ruby-doc.org/core/Array.html
    Array::[]	A			Ruby class methods							Returns a new array populated with the given objects.	http://ruby-doc.org/core/Array.html#method-c-5B-5D
    Array::new	A			Ruby class methods							<pre><code>new(size=0, default=nil)\nnew(array)\nnew(size) {|index| block }</code></pre>Returns a new array.	http://ruby-doc.org/core/Array.html#method-c-new
    Array::try_convert	A			Ruby class methods							<pre><code>try_convert(obj) → array or nil</code></pre>Tries to convert obj into an array, using to_ary method. Returns the converted array or nil if obj cannot be converted for any reason. This method can be used to check if an argument is an array.	http://ruby-doc.org/core/Array.html#method-c-try_convert
    Array#&	A			Ruby instance methods							<pre><code>ary & other_ary → new_ary</code></pre>Set Intersection — Returns a new array containing elements common to the two arrays, excluding any duplicates. The order is preserved from the original array.	http://ruby-doc.org/core/Array.html#method-i-26
    Array#*	A			Ruby instance methods							<pre><code>ary * int → new_ary\nary * str → new_string</code></pre>Repetition — With a String argument, equivalent to ary.join(str).	http://ruby-doc.org/core/Array.html#method-i-2A
    Array#+	A			Ruby instance methods							<pre><code>ary + other_ary → new_ary</code></pre>Concatenation — Returns a new array built by concatenating the two arrays together to produce a third array.	http://ruby-doc.org/core/Array.html#method-i-2B
    Array#-	A			Ruby instance methods							<pre><code>ary - other_ary → new_ary</code></pre>Array Difference	http://ruby-doc.org/core/Array.html#method-i-2D
    Array#<<	A			Ruby instance methods							<pre><code>ary << obj → ary</code></pre>Append—Pushes the given object on to the end of this array. This expression returns the array itself, so several appends may be chained together.	http://ruby-doc.org/core/Array.html#method-i-3C-3C
    Array#<=>	A			Ruby instance methods							<pre><code>ary <=> other_ary → -1, 0, +1 or nil</code></pre>Comparison — Returns an integer (-1, 0, or +1) if this array is less than, equal to, or greater than other_ary.	http://ruby-doc.org/core/Array.html#method-i-3C-3D-3E

### IA page

https://duck.co/ia/view/ruby

daxtheduck commented:
@kotoshenya Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

jdorweiler commented:
@kotoshenya this looks great! Nice use of the category pages. You can test it out on beta. 
https://beta.duckduckgo.com/?q=ruby+array%3A%3Anew&t=hb&ia=about

jdorweiler commented:
Can you remove "Ruby" from the class names though. So "Ruby class methods" should be just "class methods". That will help things like this trigger properly. https://beta.duckduckgo.com/?q=ruby+instance+methods&ia=iatesting

jdorweiler commented:
Could you get the code block for things like this?
![selection_090](https://cloud.githubusercontent.com/assets/1882892/17216837/449ff390-54b0-11e6-8971-20da34652216.png)
![selection_089](https://cloud.githubusercontent.com/assets/1882892/17216844/493dd6ec-54b0-11e6-9789-8e3eca2074fb.png)

kotoshenya commented:
The beta server doesn’t seem to have picked up those commits, but this is what [Ruby Array#length](https://beta.duckduckgo.com/?q=Ruby+Array%23length&ia=about) should look like now:

![screen shot 2016-07-29 at 00 39 30](https://cloud.githubusercontent.com/assets/6856120/17233126/1ef9975c-5525-11e6-81f5-45a5f700166d.png)

And a [longer example](https://beta.duckduckgo.com/?q=ruby+Kernel%23sprintf&ia=about):

![screen shot 2016-07-29 at 00 46 44](https://cloud.githubusercontent.com/assets/6856120/17233224/0ea73084-5526-11e6-9ef2-0c398404ca6f.png)

I’m not doing anything to escape `<`s and `&`s in output, should I be replacing them with entities?

Also, I’ve used `<br>`s to separate paragraphs, would `<p>`s render as well?

tagawa commented:
@kotoshenya Thank you for making this and adding the latest updates. Changes are not installed on the beta server automatically but I think @jdorweiler will be able to do it later.

jdorweiler commented:
@kotoshenya looking good. Just a few more comments and it should be ready to go.

- For the longer example, can you limit it to ~3 sentences.
- escape any hex characters with `\`. So `\x20` should be `\\x20'
- We can make the triggering on this much better by adding redirects. Could you add redirects for entries that have a `#` or `::` in their name?

For example, a redirect for `Array#length` would look like this:
```
Array length    R    Array#length
```
This means the query "array length" will now trigger and show the results for "Array#length". Same idea for things like `Regex::compile`.

kotoshenya commented:
> escape any hex characters with \. So \x20 should be `\x20'

@jdorweiler is that needed for all non-ASCII characters, or those which are invalid in JSON strings?

jdorweiler commented:
Just the hex characters. One of our processing scripts to install the fathead doesn't like them.

kotoshenya commented:
I’m guessing that the problem was unescaped `\`s; literal strings like `"\n"` were also appearing in the output which was probably messing up the layout.  I’ve escaped everything so `\`s are either followed by another `\` or `n`.  There are still literal Unicode characters like `→`, which appear to render fine.

jdorweiler commented:
You're right. The literal characters were file. This was one I had problems with but it look like escaping worked. 
![selection_008](https://cloud.githubusercontent.com/assets/1882892/17257531/7bc2edf6-5590-11e6-889e-4e7d2cd8d108.png)

I really like this category page showing all the class methods
![selection_009](https://cloud.githubusercontent.com/assets/1882892/17257546/8bf5cf54-5590-11e6-9bb8-ba8df7a9d2ea.png)

Looks great and I'm ready to merge unless you still wanted to add more category pages?

kotoshenya commented:
Just a couple of things (taking Enumerable as an example):

- [ ] Some (all?) redirects with spaces don’t seem to be working: I don’t get an IA for [Ruby Enumerable map](https://beta.duckduckgo.com/?q=Ruby+Enumerable+map&ia=about), but I do for [Ruby Enumerable.map](https://beta.duckduckgo.com/?q=Ruby+Enumerable.map&ia=about) and [Ruby Enumerable#map](https://beta.duckduckgo.com/?q=Ruby+Enumerable%23map&ia=about).
- [ ] I’d like to provide links from [Enumerable](https://beta.duckduckgo.com/?q=ruby+Enumerable&ia=about) to [Enumerable methods](https://beta.duckduckgo.com/184/c/Enumerable_methods?ia=list), and from [Ruby Enumerable#map](https://beta.duckduckgo.com/?q=Ruby+Enumerable%23map&ia=about) to [Enumerable](https://beta.duckduckgo.com/?q=ruby+Enumerable&ia=about).  Is that possible?

I’d also like to parse out the cross references, aliases and “see also”s from the descriptions to improve linking.  I’ve got a feeling that will be a bit more work so I can do that as a separate PR, unless it makes sense from a deployment point of view to merge everything at once.

tagawa commented:
@kotoshenya I'd say other improvements such as parsing out cross references would be fine as separate PRs, as there may be others happy to take up some of the work. The same for fixing the redirects with spaces as well perhaps. @jdorweiler can have the final say once the latest update is deployed to beta.

jdorweiler commented:
@kotoshenya 
>    Some redirects with spaces don’t seem to be working: I don’t get an IA for Ruby Enumerable map, but I do for Ruby Enumerable.map and Ruby Enumerable#map.

Let me check to make sure this isn't on our end.

>I’d also like to parse out the cross references, aliases and “see also”s from the descriptions to improve linking. I’ve got a feeling that will be a bit more work so I can do that as a separate PR, unless it makes sense from a deployment point of view to merge everything at once?

A separate PR sounds good. I think this fathead is looking great already so I'd like to see it live as soon as we can.

jdorweiler commented:
@kotoshenya it looks like the problem is on our end. The space in the search term is being removed for some reason. I'll make an issue in our internal bug tracker to fix it.  
https://beta.duckduckgo.com/?q=ruby+Enumerablecount&ia=about

I noticed  a few weird entries in "enumerable methods"
![selection_010](https://cloud.githubusercontent.com/assets/1882892/17299239/8561c556-57db-11e6-98f7-3afbb8a869c3.png)

kotoshenya commented:
[Array methods](https://beta.duckduckgo.com/?q=ruby+Array+methods&t=hd&ia=list&iax=1) look odd too:

![ruby array methods](https://cloud.githubusercontent.com/assets/6856120/17299822/be7eb1e2-5807-11e6-8235-1e3c29cdd389.png)

Those entries look fine in `output.txt` as far as I can tell, I’d guess that it’s a problem with `[]` brackets, possibly getting interpreted as link markup.

@jdorweiler I’ve escaped brackets as `\[\]` to see if that fixes the issue, though I’m really just guessing!

tagawa commented:
Just updated the beta deployment: https://beta.duckduckgo.com/?q=ruby+array+methods&ia=list&iax=1

<img width="400" alt="ruby-array-methods" src="https://cloud.githubusercontent.com/assets/94173/17313615/1a01e83c-5898-11e6-8457-e7a8627ddda1.png">

kotoshenya commented:
@tagawa That looks the same 😕 and queries like [ruby Array#[]](https://beta.duckduckgo.com/?q=ruby+Array%23%5B%5D&ia=iatesting) no longer trigger, so I think the first column at least needs to be verbatim. Also, 8aa3b0c doesn’t seem to have had an effect: related links don’t appear.

I’ve reverted the previous two commits as I didn’t really understand what I was doing, and don’t want to keep poking at it.  I’ll need some hints on how special characters in [`output.txt`](../blob/8aa3b0ca5333e03e04fb47f6965c5a7782c89c76/lib/fathead/ruby/output.txt) are processed before I can fix `parse.rb`.

jdorweiler commented:
@kotoshenya I'm still looking into the `Array#[]` problem but it's on my end for sure. Probably something to do with category and related topics also being tagged with brackets.

jdorweiler commented:
@kotoshenya I made some progress on the entries that weren't triggering.
If you switch the article entry to the key containing the spaces and make the other a redirect it all works. 

For example, here is the updated entry for `Enumerable count`

```
Enumerable count    A    ...........
Enumerable.count    R    Enumerable count
Enumerable#count   R    Enumerable count
```

kotoshenya commented:
@jdorweiler I’ve switched the keys around, so most entries should work.  I’ve listed 30 entries in ac83a14 where the class method (`<Class>::<method>`) has the same name as the instance method (`<Class>#<method>`) so one will be shadowed and removed during deduplication.

I think missing out these entries is probably fine for now, as all keywords will trigger and there’ll be a link to roughly the right part of the docs.  I can raise this as a separate issue/PR unless you think it should be fixed now.

jdorweiler commented:
@kotoshenya It works now :cake: 
https://beta.duckduckgo.com/?q=ruby+enumerable+count&ia=about&iax=1

One more thing to do is take care of the messed up tiles on the category page view
![selection_016](https://cloud.githubusercontent.com/assets/1882892/17366532/b6076c04-5959-11e6-8cd0-b6e4dd60cd48.png)

The array brackets in those code blocks seem to be causing some problems in our scripts that process your output.txt file. I think the short term fix is to just skip adding a category for any entry that contains brackets in the abstract.

I'm going to spend some time this week looking into the bug but I don't want to you to have to wait for me. Once I find and fix the bug we can add those entries back to the category view in another PR.

jdorweiler commented:
Here are some stats on your data just for fun
```
Total pages      4114
Redirects        2514    61.11%
Disambigs        0       0.00%
Categories       78      1.90%
Abstracts        4036    98.10%
```

kotoshenya commented on lib/fathead/ruby/parse.rb:
```diff
@@ -60,6 +60,10 @@ def to_s
 
+  def breaks_category_page?
+    title.include?('[') || abstract.include?('[')
+  end
```
This flags up articles containing brackets, which currently breaks the category pages (see https://github.com/duckduckgo/zeroclickinfo-fathead/pull/277#issuecomment-237232714), and should be removed when that bug is fixed.

jdorweiler replied:
:+1:  thanks, that makes it easy to revert later.

jdorweiler commented:
@kotoshenya this looks great. Probably one of the best fatheads we have now :smile: 
I'll let you know when it goes live.

moollaza commented on lib/fathead/ruby/parse.rb:
```diff
@@ -0,0 +1,173 @@
+    OutputRow.new do |row|
+      row.title = title
+      row.type = 'A' # article
+      row.categories = categories.join("\n")
```
@kotoshenya @jdorweiler I believe this actually needs to be `\\n`?

Same for the line below.

kotoshenya replied:
@moollaza Literal `\n`s are all escaped as `\\n` on [line 30](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/277/files/db1be20bdd8dd5f2fb470160c6e41808567eae18#diff-e5575cf2178993d658969bb2df779a74R30) when the row is printed.

That’s perhaps not obvious from the way the code is organized, though; I’ll have a think about how I can make it clearer.

moollaza replied:
@kotoshenya looking at the output.txt though, I'm seeing the categories are joined with `\n` (instead of `\\n`) so the result is that only the last category is being recognized by our parser internally.

This causes the infobox to only link to one category instead of multiple.

e.g. https://duckduckgo.com/?q=ruby+string+hex&ia=about

Should be linking to two categories: `Ruby instance methods` and `Ruby String methods`

kotoshenya replied:
Ah! I understand now. Backslash escaping is hard :wink:

jdorweiler replied:
Not escaping the new line is okay. There's an upper limit of 750 items in a category though. It looks like `Ruby instance methods` is above that.

I just checked this by removing enough entries to bring it below the category page limit. 
![selection_020](https://cloud.githubusercontent.com/assets/1882892/17782334/fc166a18-6540-11e6-90b3-401e47ad825e.png)
